### ISSUE TITLE
Libtorrent DHT nodes have polluted routing tables

### ISSUE THREAD

**Issue by the8472** [arvidn/libtorrent#792]
Raw observations can be found in [this gist](https://gist.github.com/the8472/eb7afdd2373c858ae17676b06143f39b). It's basically a dump of a lookup and the `nodes` entries and their respective sources.

[The lines marked with `!`](https://gist.github.com/the8472/eb7afdd2373c858ae17676b06143f39b#file-gistfile1-txt-L20-L26) are the interesting ones.
They're the ones rejected by my sanitizer that tracks IDs in `node` lists vs. `id` fields in responses across lookups.

Anyway, you can see that some of those come from a libtorrent node. In particular its home bucket. 

The attackers dynamically spoof node IDs either based on the target of lookups  or on the requester's ID. My guess would be that the request with something close to the requestee's ID for `find_node` and something close to the infohash for `get_peers`.

Anyway, they're not just getting into one home bucket but into lots of them. It's probably some variation on the _horizontal attack_ from [Real-World Sybil Attacks in BitTorrent Mainline DHT](https://www.cl.cam.ac.uk/~lw525/publications/security.pdf).

It does not induce catastrophic failures since most of those nodes seem to stay silent and don't spew back even more bogus nodes (probably just harvesting data) but at the very least it degrades lookup performance and possibly leads to incorrect announces from weak implementations that have lax lookup algorithms or something like that.

Defenses against this are possible. My own implementation mostly manages to not acquire such entries in the first place and when it does it rids itself from them fairly swiftly. All without implementing BEP42. And BEP42 would not applicable to routing table maintenance anyway, only to announce node selection during getpeers lookups.

If there is some interest in hardening LT I would share my algorithms. They're not very complex, just some subtle gotchas that have to be considered.

**Comment by arvidn** [comment]
yeah, sure. have you documented what you do to mitigate this somewhere?

**Comment by the8472** [comment]
Not yet, wanted to gauge interest first. Will do so in the next few days.

**Comment by the8472** [comment]
https://github.com/the8472/mldht/blob/sanitizing-docs/docs/sanitizing-algorithms.rst

The routing table part will be the most interesting for you. Inefficient lookups are not a problem for everyone, only for the executing node.

Feel free to ask questions.

**Comment by aldenml** [comment]
Well, that's a long list of very good measures IMO. It would be great to have them implemented in RC_1_1. I'm not sure the current architecture allows for a comprehensive implementation of these ideas (just a comment).

**Comment by the8472** [comment]
*poke* @arvidn 

Is the linked doc useful?

**Comment by the8472** [comment]
2018, still an issue:

LT1.2 on IPv4:

```
2FC0D158 DC2959EA CD43992A 3C3DF952 FDFC0721 000003E1 1DEBDE89 A9D80539 B7D6A794 4AD780D4 163.172.182.147:8420    !   fail:0 src:1 call:0 rsp:0 acc:0 [2FC0D158 DC2959EA CD43992A 3C3DF91F 1F3B1D38//46.146.165.103:6881;seen:PT7.074S;age:PT7.074S;ver:0x4C540102/LT��]
2FC0D158 DC2959EA CD43992A 3C3DF91F 1F3B1D38 000003E1 1DEBDE89 A9D80539 B7D6A7D9 A8109ACD  46.146.165.103:6881  🔒     fail:0 src:7 call:1 rsp:1 acc:1 [2FC0F048 B7D4E7E8 8303ECAD 2F2E9EDA FE093D71//124.171.133.111:20729;seen:PT7.441S;age:PT7.441S;ver:0x4C540101/LT��, 2FC0D7F1 F1BBE9EB B3A6DB3C 870C3E99 245E0D90//84.52.35.176:38206;seen:PT7.074S;age:PT7.074S;ver:0x5554AD6C/UT�l, 2FC0CDB2 8A03C85D A4A648DB 6BD76F81 6C376284//24.246.45.58:60436;seen:PT28.062S;age:PT28.062S;ver:0x5554AD2C/UT�,, 2FC0D665 E60EF171 2819BBBA 012F99ED F5E7120C//73.206.87.128:59696;seen:PT3.181S;age:PT3.181S;ver:0x5554A5B1/UT��, 2FC0C6E1 33201D87 172B12A6 03CC3E32 6ABC483E//80.7.42.127:56896;seen:PT2.31S;age:PT2.31S;ver:0x5554AD47/UT�G, 2FC0DF12 D4DE3A6D F42C881D 56F2B74F C3E37C95//186.81.48.132:26628;seen:PT3.181S;age:PT3.181S;ver:0x5554AD6C/UT�l, 2FC0C45C EE66EB43 72723D29 67D77D5C 386607AE//88.231.106.127:32469;seen:PT2.31S;age:PT2.31S;ver:0x5554AD6C/UT�l]
2FC0D158 DC2959EA CD4393CF 8E47287B 13FD9807 000003E1 1DEBDE89 A9D80FDC 05AC76BD A4D61FF2   142.4.212.116:6881    !   fail:0 src:1 call:0 rsp:0 acc:0 [2FC0D158 DC2959EA CD43992A 3C3DF91F 1F3B1D38//46.146.165.103:6881;seen:PT7.074S;age:PT7.074S;ver:0x4C540102/LT��]
2FC0D158 DC2959EA CD43B030 FF94EA24 5CA06638 000003E1 1DEBDE89 A9D82C23 747FB4E2 EB8BE1CD 163.172.231.166:6881    !   fail:0 src:1 call:0 rsp:0 acc:0 [2FC0D158 DC2959EA CD43992A 3C3DF91F 1F3B1D38//46.146.165.103:6881;seen:PT7.074S;age:PT7.074S;ver:0x4C540102/LT��]
```

LT 1.2 on IPv6

```
2C8AF50E EAA64011 4C0FC1C2 D96178E2 57ACB34C 0001BC09 9A83CC5E 5E0D6BDD 013782C0 516A4292          [2001:41d0:a:362b:10:10:100:100]:10064   !   fail:0 src:1 call:0 rsp:0 acc:0 [2C8AF50E EAA64011 4C0F7D9B 4F9C038B 99C3F3EE//2607:fb90:a2e6:b9a2:a55b:4987:33de:d4d7:42140;sent:PT6.522S;seen:PT13.799S;age:PT7M14.509S;verified;rtt:192.73209338406662;ver:0x4C540102/LT��]
2C8AF50E EAA64011 4C0FDC4B 5791CE14 0DB6BE3D 0001BC09 9A83CC5E 5E0D7654 8FC73436 0B704FE3    [2a05:d018:2d6:c000:276e:bde:2d3:10b2]:6892    !   fail:-2 src:1 call:0 rsp:0 acc:0 [2C8AF50E EAA64011 4C0F7D9B 4F9C038B 99C3F3EE//2607:fb90:a2e6:b9a2:a55b:4987:33de:d4d7:42140;sent:PT6.522S;seen:PT13.799S;age:PT7M14.509S;verified;rtt:192.73209338406662;ver:0x4C540102/LT��]
2C8AF50E EAA64011 4C0F7D9B 4F9C038B 99C3F3EE 0001BC09 9A83CC5E 5E0DD784 97CAF9A9 9F050230 [2607:fb90:a2e6:b9a2:a55b:4987:33de:d4d7]:42140 🔒🌲    fail:0 src:4 call:1 rsp:1 acc:1 [2C8A63B3 D7CC8EC2 6C821A20 C5951428 46C8E377//2001:da8:8000:e103:0:0:0:3b78:51413;sent:PT5.85S;seen:PT13.604S;age:PT79H15M1.883S;verified;rtt:671.8471422784661, 2C871E4F 4129EB8B 4A703E83 E181AF36 4B5D4544//2001:470:6c:eaa:0:0:0:2:4443;sent:PT28.96S;seen:PT31.472S;age:PT165H13M25.104S;fail:1;verified;rtt:98.86223523320561, 2C85E146 AC365729 8447C479 3C249F37 3DE60711//2001:1af8:4700:b0d0:e400:0:0:0:44564;seen:PT20.023S;age:PT34.041S;verified;rtt:33.199999999999996, 2C86E546 72E85B22 558B1A55 40012D47 AFF81B32//2607:fb90:6508:9dae:783e:bd4a:82dc:50c2:49974;seen:PT1.795S;age:PT8.53S;verified;rtt:267.7;ver:0x4C540102/LT��]
2C8AF50E EAA64011 4C0F7D9B 4F9C03D2 4098DAC0 0001BC09 9A83CC5E 5E0DD784 97CAF9F0 465E2B1E               [2a06:8187:fb11:1:0:0:13:1]:2000    !   fail:-2 src:1 call:0 rsp:0 acc:0 [2C8AF50E EAA64011 4C0F7D9B 4F9C038B 99C3F3EE//2607:fb90:a2e6:b9a2:a55b:4987:33de:d4d7:42140;sent:PT6.522S;seen:PT13.799S;age:PT7M14.509S;verified;rtt:192.73209338406662;ver:0x4C540102/LT��]
```

Note how the spoofing nodes share around ~80 ID bits with the LT node returning the spoofed contacts. This is clearly not something that would occur naturally.

**Comment by arvidn** [comment]
@the8472 re-reading your algorithm document, I believe libtorrent does all of those things except re-verifying all nodes in the same bucket as a node with a mismatching ID has been discovered. Is that a key part of keeping the table clean?

Come to think of it, it's possible I'm also not cleaning out nodes when they report a new ID, but rather just ignore them, causing a timeout.

**Comment by the8472** [comment]
> I believe libtorrent does all of those things

Hrm, then I wonder how nodes which spoof 80 shared prefix bits to everyone they reply to get into the LT routing table while mldht manages to avoid being fooled by them.

> Is that a key part of keeping the table clean?

Key part? No. It just speeds up cleanup if something slips through the other checks.

I think the most powerful tools in the box are:

* observer-independence verification
* expected ID verification
* the strictest checks (transaction ID + reachability + exact socket address match + expected ID match) for promotions into main buckets

Hypothetically the multi-path verification from S/Kademlia should be pretty good too as extension of the expected ID verification, but I have not yet had the need to implement that.

But I have not tested them in isolation. I simply kept heaping on defensive code until the problematic entries were gone from my routing tables.

> Come to think of it, it's possible I'm also not cleaning out nodes when they report a new ID, but rather just ignore them, causing a timeout.

That could make a difference, but ideally this specific kind of attack shouldn't make it in in the first place.

**Comment by ssiloti** [comment]
Of the three you listed, the only one which is missing from libtorrent right now is observer-independence verification. Doing it via a shared routing table for multi-homed nodes would require replacing libtorrent's simplified routing table with a tree based implementation. I'm not too enthusiastic about that because most nodes are single homed so we'd need some additional mitigation anyways.

**Comment by the8472** [comment]
Right, I'm running all my v6 nodes with multiple sockets to test multi-homing.
I'll look at some single-homed IPv4 tables to see how well it fares without that.

**Comment by the8472** [comment]
Also note that in principle you can implement that with a ping from a temporary socket using a different source node ID. Although that's kind of inelegant because your temp socket might end up in someone's replacement bucket

**Comment by the8472** [comment]
On a single-homed IPv4 node It looks like multiple spoofed entries manage to get into my replacement bucket. But they do not attain verified status, which means they come from unsolicited requests.

```
A555200000000000000000000000000000000000/18  ihash:   18 s:   13 p:   15               [Home:A555260A66E41DFFA3F3ECCD2DDF4004C8853D2A]
  Entries (7)
    A5552D46 14DB24E2 8AD84E39 45A8213F F97B1388//200.150.169.110:33286;sent:PT3M59.424S;seen:PT3M59.15S;age:PT34M30.776S;verified;rtt:281.7657993725665;ver:0x5554AD6C/UT�l 
    A5552F66 B8924C04 557287DC 1CC8F5E9 ABA7631D//31.16.251.81:4722;sent:PT2M30.322S;seen:PT2M19.055S;age:PT34M30.559S;fail:8;verified;rtt:39.89725795;ver:0x55547021/UTp! 
    A5552992 E352121C D125FB75 C6875621 C5BA6AD7//176.63.26.106:9270;sent:PT3M59.425S;seen:PT3M59.379S;age:PT34M30.406S;verified;rtt:43.830947051551156;ver:0x5554AD6C/UT�l 
    A5553458 7662BBDB 549FA606 4DE66653 29423F93//184.157.84.224:51248;sent:PT50.315S;seen:PT50.148S;age:PT34M29.908S;verified;rtt:206.96112252923936;ver:0x4C540102/LT�� 
    A5553A68 AC4BCC86 DEF594B2 2ABD695E 2302868F//27.154.174.167:13385;sent:PT3M49.109S;seen:PT2M46.367S;age:PT34M29.437S;verified;rtt:421.97032268915586 
    A5552686 DDF543C7 63CBCD48 BE549DD5 63C85D32//70.49.40.252:28096;sent:PT4M10.139S;seen:PT2M50.224S;age:PT34M24.001S;fail:7;verified;rtt:122.0;ver:0x5554AD6C/UT�l 
    A5552AA5 3CBB4C13 995292B8 7D6C1321 6FF2379F//122.57.92.189:6881;sent:PT3M59.424S;seen:PT3M59.105S;age:PT6M26.735S;verified;rtt:320.4703999999999;ver:0x4C540010/LT�� 
  Replacements (7)
    A5553911 E15AA453 9032AEBE F35B80A6 5FE4B93E//62.83.166.118:39207;sent:PT3M48.754S;seen:PT20M16.5S;age:PT20M16.5S;fail:5;ver:0x5554AD6C/UT�l
    A555260A 66E41DFF A3F35E0D 2830D916 3B751DD3//205.209.186.206:6881;seen:PT5M8.827S;age:PT5M8.827S
    A555260A 66E41DFF A3F3ECCD 2DDF4001 5CC2D141//121.196.207.196:6882;seen:PT22.596S;age:PT22.596S
    A5552894 776063A6 26F81ED3 EAF34A78 C7D16398//83.202.86.79:55942;seen:PT3.447S;age:PT47.426S;ver:0x5554AB58/UT�X
    A555260A 66E41DFF A3F32F34 3BB3EE52 426805BB//195.154.79.185:6881;seen:PT2M18.599S;age:PT2M18.599S
    A555260A 66E41DFF A3F3338A 8118ED4C 9E3F2FFF//163.172.5.207:43231;seen:PT8M42.381S;age:PT8M42.381S
    A5552741 CCFC6B84 626007B2 4D2374BC 85C9CEB1//79.87.168.55:6881;seen:PT3M45.303S;age:PT3M45.303S
```

But it looks like my promotion logic is *very* lazy about actively verifying when the main bucket is not full. That might be protecting me here by accident. I'll have to do more testing.

**Comment by the8472** [comment]
For the purpose of checking this again, is the fix only in 1.2 or was it also backported to 1.1?

**Comment by arvidn** [comment]
it doesn't look like it was back-ported. but it probably should be
